### PR TITLE
Init addition of online interfaces in Covariance

### PR DESCRIPTION
**Pull request by Alexandr-Solovev**
# Description
This pr includes example and code of online interfaces for covariance and base for future online implementations.

Changes proposed in this pull request:
- Adding `partial_compute` and `finalize_compute`
- Adding  `cov_online_dense` and `cor_online_dense` example
- Implementing bazel tests for online
- Implementing CPU kernel via DAAL code
- DPCPP implementation of GPU kernel

### CODE REVIEW

File: cpp/oneapi/dal/finalize_compute.hpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        dal::detail::spmd_policy<dal::detail::data_parallel_policy>{
+            dal::detail::data_parallel_policy{ comm.get_queue() },
+            comm },
+        std::forward<Args>(args)...);
+}
+#endif
+
+} // namespace preview
+
+} // namespace oneapi::dal
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/finalize_compute.hpp:65]:
```suggestion
} // namespace oneapi::dal

```

File: cpp/oneapi/dal/detail/finalize_compute_ops.hpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+auto finalize_compute_dispatch(Head&& head, Tail&&... tail) {
+    using dispatcher_t = ops_policy_dispatcher<std::decay_t<Head>, tagged_finalize_compute_ops>;
+    return dispatcher_t{}(std::forward<Head>(head), std::forward<Tail>(tail)...);
+}
+
+} // namespace v1
+
+using v1::finalize_compute_dispatch;
+
+} // namespace oneapi::dal::detail
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/detail/finalize_compute_ops.hpp:40]:
```suggestion
} // namespace oneapi::dal::detail

```

**Review by inteldimitrius — APPROVED** [review]
Look good to me, but some checks are red.
And refer to @Vika-F it is better to make some comment-header for kernel function to explain math logic. But it is optional.

File: cpp/oneapi/dal/algo/covariance/backend/cpu/partial_compute_kernel_dense.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            interop::call_daal_kernel<Float, daal_covariance_kernel_t>(ctx,
+                                                                       daal_data.get(),
+                                                                       daal_nobs_matrix.get(),
+                                                                       daal_crossproduct.get(),
+                                                                       daal_sums.get(),
+                                                                       &daal_parameter));
+        result.set_sums(interop::convert_from_daal_homogen_table<Float>(daal_sums));
+        result.set_nobs(interop::convert_from_daal_homogen_table<Float>(daal_nobs_matrix));
+        result.set_crossproduct(interop::convert_from_daal_homogen_table<Float>(daal_crossproduct));
+    }
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/covariance/backend/cpu/partial_compute_kernel_dense.cpp:72]: Can you please separate it into 2 different functions?

File: cpp/oneapi/dal/algo/covariance/backend/gpu/partial_compute_kernel_dense_dpc.cpp [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    ONEDAL_ASSERT(data.has_data());
+
+    const std::int64_t column_count = data.get_dimension(1);
+    auto xtx = pr::ndarray<Float, 2>::empty(q, { column_count, column_count }, alloc::device);
+
+    sycl::event gemm_event;
+    {
+        ONEDAL_PROFILER_TASK(gemm, q);
+        gemm_event = gemm(q, data.t(), data, xtx, Float(1.0), Float(0.0));
+        gemm_event.wait_and_throw();
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/covariance/backend/gpu/partial_compute_kernel_dense_dpc.cpp:73]: Do we really need this `wait` call?

File: cpp/oneapi/dal/algo/covariance/backend/gpu/partial_compute_kernel_dense_dpc.cpp [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    auto crossproducts_data = crossproducts.get_data();
+    auto sums_data = sums.get_data();
+
+    auto update_event = q.submit([&](sycl::handler& cgh) {
+        const auto range = sycl::range<2>(column_count, column_count);
+
+        cgh.depends_on(deps);
+        cgh.parallel_for(range, [=](sycl::item<2> id) {
+            auto row = id[0];
+            auto col = id[1];
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/covariance/backend/gpu/partial_compute_kernel_dense_dpc.cpp:111]:
```suggestion
            const std::int64_t row = id[0], col = id[1];
```

File: cpp/oneapi/dal/algo/covariance/test/fixture.hpp [hunk 6]
```diff
@@ -46,6 +48,31 @@ class covariance_test : public te::crtp_algo_fixture<TestType, Derived> {
         return te::table_id::homogen<Float>();
     }
 
+    template <typename Float>
+    std::vector<dal::table> split_table_by_rows(const dal::table& t, std::int64_t split_count) {
+        ONEDAL_ASSERT(split_count > 0);
+        ONEDAL_ASSERT(split_count <= t.get_row_count());
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/covariance/test/fixture.hpp:54]:
```suggestion
        ONEDAL_ASSERT(0l < split_count);
        ONEDAL_ASSERT(split_count <= t.get_row_count());
```

File: cpp/oneapi/dal/algo/covariance/test/fixture.hpp [hunk 7]
```diff
@@ -46,6 +48,31 @@ class covariance_test : public te::crtp_algo_fixture<TestType, Derived> {
...
+        const std::int64_t column_count = t.get_column_count();
+        const std::int64_t block_size_regular = row_count / split_count;
+        const std::int64_t block_size_tail = row_count % split_count;
+
+        std::vector<dal::table> result(split_count);
+
+        std::int64_t row_offset = 0;
+        for (std::int64_t i = 0; i < split_count; i++) {
+            const std::int64_t tail = std::int64_t(i + 1 == split_count) * block_size_tail;
+            const std::int64_t block_size = block_size_regular + tail;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/covariance/test/fixture.hpp:66]: Please consider using [`uniform_blocking`](https://github.com/oneapi-src/oneDAL/blob/d69f5fc3f461561445ed704bbd151180a09f4218/cpp/oneapi/dal/backend/common.hpp#L155)

File: cpp/oneapi/dal/algo/covariance/test/fixture.hpp [hunk 8]
```diff
@@ -107,6 +134,102 @@ class covariance_test : public te::crtp_algo_fixture<TestType, Derived> {
...
+        INFO("create descriptor cov")
+        cov_desc =
+            covariance::descriptor<Float, Method, covariance::task::compute>().set_result_options(
+                covariance::result_options::cov_matrix);
+        INFO("run compute optional: cov");
+        partial_result = dal::covariance::partial_compute_result();
+        input_table = split_table_by_rows<double>(data, 10);
+        for (int i = 0; i < 10; i++) {
+            partial_result = this->partial_compute(cov_desc, partial_result, input_table[i]);
+        }
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/covariance/test/fixture.hpp:163]:
```suggestion
        for (std::int64_t i = 0; i < 10; ++i) {
            partial_result = this->partial_compute(cov_desc, partial_result, input_table[i]);
        }
```

**Review by KulikovNikita — APPROVED** [review]
Generally - looks fine to me. Thx!

**Pull request merged**